src/pattern-acceptor.js

```diff
@@ -0,0 +1,532 @@
+/**
+ * Copyright 2018 Shape Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-use-before-define */
+
+const syntaxCharacters = '^$\\.*+?()[]{}|'.split('');
+const extendedSyntaxCharacters = '^$\\.*+?()[|'.split('');
+
+const controlEscapeCharacters = 'fnrtv'.split('');
+const controlEscapeCharacterValues = { 'f': '\f'.charCodeAt(0), 'n': '\n'.charCodeAt(0), 'r': '\r'.charCodeAt(0), 't': '\t'.charCodeAt(0), 'v': '\v'.charCodeAt(0) };
+
+const controlCharacters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
+const hexDigits = '0123456789abcdefABCDEF'.split('');
+const decimalDigits = '0123456789'.split('');
+const octalDigits = '01234567'.split('');
+
+
+class PatternAcceptorState {
+  constructor(pattern, unicode) {
+    this.pattern = pattern;
+    this.unicode = unicode;
+    this.index = 0;
+    this.backreferences = [];
+    this.capturingGroups = 0;
+  }
+
+  empty() {
+    return this.index >= this.pattern.length;
+  }
+
+  nextCodePoint() {
+    if (this.empty()) {
+      return false;
```

Comment:
I'd still prefer to use `null` rather than `false` for "not found" in methods which can return a string, i.e. here and in `eatAny` (and `eatNaturalNumber` after my comment below).